### PR TITLE
net: lib: download_client: Non blocking connect 

### PR DESCRIPTION
**Pull request by SeppoTakalo**
### Refactor API for non-blocking support
    
* Don't block on download_client_connect() or download_client_disconnect()
* Refactor download_client_connect() to download_client_set_host()
  so the name indicates that it does not yet initiate anything.
* Add function download_client_get() that combines set_host,
  start and disconnect.
* Do all name queries and socket connections in the background
  thread.
* Add new event `DOWNLOAD_CLIENT_EVT_CLOSED` to indicate that background thread has closed the connection and new download can be started.
    
Now there are two options to use this API:
1. set_host() and call start() for multiple files,
   and finally call disconnect().
2. Call get().
    
On both cases, init() should only be called once.
* Both cycles can be repeated.
* Both use cases can be stopped/aborted by disconnect().

### Refactor the handler loop 

* Add mutexes to protect internal state
* Refactor the handling loop so there is no "goto". This should lead to more maintainable code. I had hard time understanding the request-loop and its many corner cases. It should be clearer now.
* Various minor fixes

### Remove pause and resume
    
Pause and resume are non thread safe operations and
there are no use for them as we can just disconnect
and resume with offset.

### Various fixes

* Return -EALREADY if connect called twice
* download_client_disconnect() should be called before connecting again.
* Fix unranged HTTP GET
* Fix CoAP download ending
   For GET requests where we don't know the full size, we must
   use the last_block marker and end the transfer at that point.

### CODE REVIEW

File: subsys/net/lib/download_client/src/download_client.c [hunk 1]
```diff
@@ -483,6 +496,21 @@ void download_thread(void *client, void *a, void *b)
...
+
+		if (rc == 0) {
+			if (IS_ENABLED(CONFIG_COAP) &&
+			    (dl->proto == IPPROTO_UDP || dl->proto == IPPROTO_DTLS_1_2)) {
+				coap_block_init(client, dl->progress);
+			}
+			rc = request_send(dl);
+		}
+		if (rc) {
+			error_evt_send(dl, rc);
```
Comment (lucasvdp) [inline subsys/net/lib/download_client/src/download_client.c:510]: Looks like `rc` can be both a positive error code from the socket operations and a negative error code we generate ourselves. `error_evt_send()` will assert on negative error codes.
It would be useful to know why we failed to determine if we could try again.

File: subsys/net/lib/download_client/src/download_client.c [hunk 2]
```diff
@@ -759,17 +764,6 @@ int download_client_start(struct download_client *client, const char *file,
...
-			coap_block_init(client, from);
-		}
-	}
-
-	err = request_send(client);
-	if (err) {
-		return err;
-	}
-
 	LOG_INF("Downloading: %s [%u]", client->file, client->progress);
```
Comment (lucasvdp) [inline subsys/net/lib/download_client/src/download_client.c:767]: Maybe move the log line as well?

File: subsys/net/lib/download_client/src/download_client.c [hunk 3]
```diff
@@ -680,13 +708,11 @@ int download_client_init(struct download_client *const client,
 int download_client_connect(struct download_client *client, const char *host,
```
Comment (lucasvdp) [inline subsys/net/lib/download_client/src/download_client.c:708]: Does it make still make sense to have separate `_connect()` and `_start()` functions? Or maybe we can rename this function?

File: subsys/net/lib/download_client/src/download_client.c (hunk 1 above)
Comment (SeppoTakalo) [reply subsys/net/lib/download_client/src/download_client.c:510]: Where do we get positive values?
`http_request_send()`, `coap_request_send()`  and `socket_send()` all seem to `return 0;` when things work. Positive values from `send()` are eaten in the `socket_send()`.

Comment (lucasvdp) [reply subsys/net/lib/download_client/src/download_client.c:510]: I don't remember, but it was probably from connect. I think I tried a non existing URI, probably by changing the port number. After the timeout I got the Assert.

Comment (SeppoTakalo) [reply subsys/net/lib/download_client/src/download_client.c:510]: `connect()` can only return 0 or -1.

I cannot spot any errors, I tried:
* Correct HTTP address
* Correct HTTPS address
* Incorrect HTTP and HTTPS addres
* Non existent domain

```
*** Booting Zephyr OS build v3.2.99-ncs2-33-ge414f0842616 ***
Download client sample started
Waiting for network.. OK
Downloading http://speedtest.ftp.otenet.gr/files/test100k.db
[ 100% ] |==================================================| (102400/102400 bytes)
Download completed in 9573 ms @ 10696 bytes per sec, total 102400 bytes
Bye
*** Booting Zephyr OS build v3.2.99-ncs2-33-ge414f0842616 ***
Download client sample started
Provisioning certificate
Waiting for network.. OK
Downloading https://nrfconnectsdk.s3.eu-central-1.amazonaws.com/sample-img-100kb.png
[ 100% ] |==================================================| (102923/102923 bytes)
Download completed in 41672 ms @ 2469 bytes per sec, total 102923 bytes
Bye
*** Booting Zephyr OS build v3.2.99-ncs2-33-ge414f0842616 ***
Download client sample started
Waiting for network.. OK
Downloading http://example.com:1234/nonexistent
Error 116 during download
*** Booting Zephyr OS build v3.2.99-ncs2-33-ge414f0842616 ***
Download client sample started
Certificate match
Waiting for network.. OK
Downloading https://example.com:1234/nonexistent
Error 116 during download
*** Booting Zephyr OS build v3.2.99-ncs2-33-ge414f0842616 ***
Download client sample started
Certificate match
Waiting for network.. OK
Downloading http://nonexistend.domain/nonexistent
Error 118 during download

```

All usecases seem to return sensible errors.

File: subsys/net/lib/download_client/src/download_client.c (hunk 3 above)
Comment (SeppoTakalo) [reply subsys/net/lib/download_client/src/download_client.c:708]: I did not plan to do any API changes.

File: subsys/net/lib/download_client/src/download_client.c (hunk 2 above)
Comment (SeppoTakalo) [reply subsys/net/lib/download_client/src/download_client.c:767]: This does not matter. It kind of starts the request here.

**Comment by SeppoTakalo** [comment]
I managed to find another bug in the CoAP implementation.
Basically it did not end the transfer when we received a last block of a transfer where we don't know the full size.

It should be fixed now.

File: subsys/net/lib/download_client/src/download_client.c (hunk 1 above)
Comment (SeppoTakalo) [reply subsys/net/lib/download_client/src/download_client.c]: Oh.. now I see.
I overlooked this, as I was claiming that none of the mentioned functions return positive erros, I did not notice that Lucas already told me that `error_evt_send()` expects positive. In my mind I was thinking opposite.

Now it is fixed. I convert the error code to positive for the event handler, which in turn convers it to negative..

**Comment by SeppoTakalo** [comment]
I think we should do the API change on a separate PR once this goes through.
Or at least that is my proposal.

**Comment by SeppoTakalo** [comment]
What would be the correct API for the future:
* `init()`, `setup()` same as connect, and `start()`
* `init()`, `fetch()` that combines connect and start.

**Comment by SeppoTakalo** [comment]
The API has been refactored now as proposed by others.
Please review again.

File: subsys/net/lib/download_client/src/download_client.c
Comment (lucasvdp) [inline subsys/net/lib/download_client/src/download_client.c:1]: `download_client_disconnect()` is still blocking. The `close()` function seems to wait for `connect()` to succeed or fail.

File: include/net/download_client.h [hunk 4]
```diff
@@ -266,6 +287,34 @@ int download_client_file_size_get(struct download_client *client, size_t *size);
...
+ * @kconfig{CONFIG_DOWNLOAD_CLIENT_COAP_BLOCK_SIZE} bytes for CoAP,
+ * which are delivered to the application
+ * via @ref DOWNLOAD_CLIENT_EVT_FRAGMENT events.
+ *
+ * @param[in] client	Client instance.
+ * @param[in] host	Name of the host to connect to, null-terminated.
+ *			Can include scheme and port number, defaults to
+ *			HTTP or HTTPS if no scheme is provided.
+ * @param[in] config	Configuration options.
+ * @param[in] file	File to download, null-terminated.
```
Comment (lucasvdp) [inline include/net/download_client.h:309]: No need for a separate host and file. You can use the complete URI for both the server and the filename.

File: subsys/net/lib/download_client/src/download_client.c [hunk 5]
```diff
@@ -742,6 +756,9 @@ int download_client_disconnect(struct download_client *const client)
 	}
 
 	client->fd = -1;
+	client->close_when_done = false;
+	client->host = NULL;
```
Comment (lucasvdp) [inline subsys/net/lib/download_client/src/download_client.c:760]: I got an assert on http.c:57 because of this. We have to be carefull changing this variables from multiple threads.

File: include/net/download_client.h (hunk 4 above)
Comment (SeppoTakalo) [reply include/net/download_client.h:309]: Well, many of the existing FOTA APIs separate file and host, so why change it here?

File: subsys/net/lib/download_client/src/download_client.c (hunk 5 above)
Comment (SeppoTakalo) [reply subsys/net/lib/download_client/src/download_client.c:760]: This looks like we actually need to add mutex to protect the content of client structure.

**Comment by SeppoTakalo** [comment]
Seeking for a opinion here:
It looks like refactoring the download client to be a non-blocking, we introduce issues with thread safety. Those issues most probably are there already, but not so easily triggered.
If I add mutexes to protect the internal data structure, I end up dead-blocking in pause and resume functions. Instead of fixing those, I'm proposing to remove pause and resume functionality for the reason that applications should instead cancel the download by calling `download_client_disconnect()` and then resume it with `_start(..., from);` or `_get(..., from)`. Another reason for removing the pause is that I don't think it is safe to suspend a thread that is holding a socket handle. I unaware of any tests on that area. While suspended, the sockets might be closed and another opened which might might turn the socket handle to actually point to entirely different socket, after all it is a integer and not a OS provided file handle.

So what are your opinion @lemrey, @lucasvdp  and @rlubos ?
Also @chmielewskiandreas Since you are editing tests for the download client in https://github.com/nrfconnect/sdk-nrf/pull/10343, can you provide your opinion here?

**Comment by rlubos** [comment]
> Then _disconnect() is still the same, and it directly touched the connected socket.

Ok, I stand corrected. Still, it's just another reason to re-think the validity of pause/resume (at least in current form).

> Or, another option would be that I refactor disconnect() to also be just a flag/command that request background thread to do the operation.

From my POV, such an approach would be a much cleaner design.

File: include/net/download_client.h (hunk 4 above)
Comment (lucasvdp) [reply include/net/download_client.h:309]: The download client can already separate the host name from the file name for you. So no need to suggest that the user also needs to do this.
If they already have one (or more) file name(s) and the host name separate they can still use the `_set_host()` (multiple) `_start()` `_disconnect()` scheme.
A function like wget also does not require you to split the uri yourself.

**Comment by lucasvdp** [comment]
The pause option is not currently used in any of the samples so I am oké with removing it.

I would vote for setting a flag/event in the `_disconnect()` function. This way you should not need a mutex and the `_disconnect()` can return immediately. Only disadvantage is that you cannot immediately retry with a new download.
I would suggest that retries should not be done immediately anyway, so with a line of explanation in the documentation I am fine with this.

**Comment by SeppoTakalo** [comment]
I ended up refactoring the main loop for quite a bit. Mainly because it used lots of `goto` commands to jump back and forth. It was hard to understand, and I'm still not 100% sure that I did cover all the corner cases. This builds and runs now, but I would like someone to properly test that I did not break anything serious

File: include/net/download_client.h [hunk 6]
```diff
@@ -182,11 +191,17 @@ struct download_client {
...
 	bool set_native_tls;
+
+	/** Close the socket when finished */
+	bool close_when_done;
 };
 
 /**
  * @brief Initialize the download client.
  *
+ * This function can only be called once per client instance as
```
Comment (peknis) [inline include/net/download_client.h:202]:
```suggestion
 * This function can only be called once in each client instance as
```

File: include/net/download_client.h (hunk 4 above)
Comment (SeppoTakalo) [reply include/net/download_client.h]: Fixed. This accepts both forms now.

**Review by peknis — APPROVED** [review]

File: subsys/net/lib/download_client/src/download_client.c [hunk 7]
```diff
@@ -745,32 +832,50 @@ int download_client_disconnect(struct download_client *const client)
...
+	}
+
+	k_mutex_lock(&client->mutex, K_FOREVER);
+
+	if (client->fd < 0) {
+		k_mutex_unlock(&client->mutex);
 		return -EINVAL;
 	}
 
 	err = close(client->fd);
```
Comment (lucasvdp) [inline subsys/net/lib/download_client/src/download_client.c:846]: The restructuring is a nice improvement over the old goto-structure. But I still don't like having the `close()` in the disconnect function. I would prefer to move all socket operations including the `close()` to the download_client thread. Putting the mutex around it will even increase the change of locking a thread.
If we just set a disconnect flag in this function, and do everything else in the download_client thread. We might not even need the mutex in this location at all. It would be good to add an event to indicate to the user that the disconnecting is finished. This to prevent them from having to retry the next `download_client_set_host()` until we no longer return `-EALREADY` (or reuse the ERROR or DONE event?).

I don't have a lot of experience with mutexes apart from them being the cause of deadlocks. I would prefer to keep the mutexes as small as possible and only around the variable you are trying to protect, not around functions calls or every element of the struct. The only elements set from outside the thread that might need protection are the URI, config, filename, offset and the disconnect_flag. Do we even need to protect these with a mutex? The worst thing that can happen is someone doing a second download request and us mixing up the file and offset.

Comment (lemrey) [reply subsys/net/lib/download_client/src/download_client.c:846]: It makes sense to move `close()` in the `download_thread` to avoid it happening while the `download_thread` is in a `connect()` call, which could be troublesome.

Comment (SeppoTakalo) [reply subsys/net/lib/download_client/src/download_client.c:846]: Thanks for the proposal.

I'll start implementing the non-blocking disconnect as well.

**Comment by SeppoTakalo** [comment]
```
[  67% ] |=================================                 | (69632/102923 bytes)Error -104 during download
[2023-04-20T10:16:22.808Z] nrf9160_common_uart.py - [33mWARNING[0m - Timeout while waiting for 'Download completed'.
```

Uhh.... harsh..
Tests fail because of performance.

**Comment by SeppoTakalo** [comment]
Both CI failures seems unrelated.
One has `rest_client` failing. Other fails to clone the repositories.

**Review by VTPeltoketo — APPROVED** [review]

File: subsys/net/lib/fota_download/src/fota_download.c [hunk 8]
```diff
@@ -383,66 +404,88 @@ int fota_download_start_with_image_type(const char *host, const char *file,
...
 	}
 
-	int err;
-
 	callback = client_callback;
 
+	if (initialized) {
+		return 0;
+	}
+
```
Comment (VTPeltoketo) [inline subsys/net/lib/fota_download/src/fota_download.c:488]: Is this correct? A local variable "initialized" is not set and this is the first usage of it.

Comment (SeppoTakalo) [reply subsys/net/lib/fota_download/src/fota_download.c:488]: Yes, it is correct.

[C11 standard](https://www.open-std.org/jtc1/sc22/wg14/www/docs/n1548.pdf) Says in **6.7.9 Initialization** clause 10:
```
 If an object that has static storage duration is not initialized explicitly,
then:
— if it has pointer type, it is initialized to a null pointer;
— if it has arithmetic type, it is initialized to (positive or unsigned) zero;
```
(This same applies to previous versions as well, at least C99 is the same)

Initializing static variables to zero, or false, will generate an [error with checkpatch](https://github.com/zephyrproject-rtos/zephyr/blob/325e35036bac56f6e21e2adcdf37d9fd616a9446/scripts/checkpatch.pl#L4050-L4058).

**Comment by rlubos** [comment]
Seems that Jenkins did not update all GH statuses... For anyone's information, twister is passing for this PR (at least at the current moment).

**Review by rlubos — APPROVED** [review]
LGTM

**Review by jayteemo — APPROVED** [review]

**Comment by SeppoTakalo** [comment]
OK. Now the CI seems broken:
```
[2023-05-02T07:22:19.095Z] Install all dependencies for Matter by invoking: pip3 install -r ./scripts/setup/requirements.nrfconnect.txt from the Matter root directory.

[2023-05-02T07:22:19.095Z] [18/835] Generating factory data HEX file...

[2023-05-02T07:22:19.095Z] FAILED: zephyr/factory_data.hex
```

**Review by MarkusLassila — APPROVED** [review]

**Comment by SeppoTakalo** [comment]
Rebased and fixed the conflict.

File: subsys/net/lib/fota_download/src/fota_download.c [hunk 9]
```diff
@@ -138,75 +136,82 @@ static int download_client_callback(const struct download_client_evt *event)
...
 		if (err != 0) {
 			LOG_ERR("dfu_target_write error %d", err);
-			int res = dfu_target_done(false);
-
-			if (res != 0) {
-				LOG_ERR("Unable to free DFU target resources");
-			}
-			first_fragment = true;
-			(void) download_client_disconnect(&dlc);
-			send_error_evt(FOTA_DOWNLOAD_ERROR_CAUSE_INVALID_UPDATE);
```
Comment (jayteemo) [inline subsys/net/lib/fota_download/src/fota_download.c:186]: what is the reason for changing the error cause here?
`FOTA_DOWNLOAD_ERROR_CAUSE_INVALID_UPDATE` seems more descriptive and appropriate.
the CI is failing because the test/fota sample expect the error code to be `INVALID_UPDATE` when an invalid modem delta update is attempted.

**Comment by jayteemo** [comment]
added `DNM`.
the `fota_download` changes break the `nrf_cloud_rest_fota` sample and also `nrf_cloud_fota`... 
fixed [in this branch](https://github.com/jayteemo/fw-nrfconnect-nrf/tree/rest_fota_fix).

it is probably worth checking the other libs/samples that use `fota_download` for similar issues.
remove the tag as you please.

File: subsys/net/lib/fota_download/src/fota_download.c (hunk 9 above)
Comment (SeppoTakalo) [reply subsys/net/lib/fota_download/src/fota_download.c:186]: That is a valid question.
It looks like this is another undocumented feature. Modem Delta update responds `-EINVAL` when it rejects the delta image. All other cases, it is just the flash write have failed. I'll add a if-else for the corner case.

**Review by stig-bjorlykke — CHANGES_REQUESTED** [review]

File: lib/bin/lwm2m_carrier/os/lwm2m_os.c [hunk 10]
```diff
@@ -512,7 +512,7 @@ int lwm2m_os_download_connect(const char *host, const struct lwm2m_os_download_c
 		.pdn_id = cfg->pdn_id,
 	};
 
-	return download_client_connect(&http_downloader, host, &config);
+	return download_client_set_host(&http_downloader, host, &config);
```
[the remark this answers is not in the document]
Comment (stig-bjorlykke) [reply lib/bin/lwm2m_carrier/os/lwm2m_os.c:515]: We will handle this in the carrier library after merge.

**Review by stig-bjorlykke — APPROVED** [review]

**Review by nordic-hani — APPROVED** [review]

File: doc/nrf/releases/release-notes-changelog.rst [hunk 11]
```diff
@@ -538,6 +538,12 @@ Libraries for networking
 
   * Removed the old API ``lwm2m_firmware_get_update_state_cb()``.
 
+* :ref:`lib_download_client` library:
+
+  * Refactored the :c:func:`download_client_connect` function to :c:func:`download_client_set_host` and made it non-blocking.
+  * Added the :c:func:`download_client_get` function that combines the functionality of functions :c:func:`download_client_set_host`, :c:func:`download_client_start`, and :c:func:`download_client_disconnect`.
+  * Removed functions :c:func:`donwload_client_pause` and :c:func:`donwload_client_resume`.
```
[the remark this answers is not in the document]
Comment (stig-bjorlykke) [reply doc/nrf/releases/release-notes-changelog.rst:545]: I can fix this in #10882

**Comment by SeppoTakalo** [comment]
Removed the DNM tag as the cloud test was passing last night.
Issue pointed out by Justin was fixed yesterday.
